src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,122 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json.Serialization.Policies;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class DefaultImmutableConverter : JsonEnumerableConverter
+    {
+        public const string ImmutableNamespace = "System.Collections.Immutable";
+
+        private const string ImmutableListGenericTypeName = "System.Collections.Immutable.ImmutableList`1";
+        private const string ImmutableListGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableList`1";
+
+        private const string ImmutableArrayGenericTypeName = "System.Collections.Immutable.ImmutableArray`1";
+
+        private const string ImmutableStackGenericTypeName = "System.Collections.Immutable.ImmutableStack`1";
+        private const string ImmutableStackGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableStack`1";
+
+        private const string ImmutableQueueGenericTypeName = "System.Collections.Immutable.ImmutableQueue`1";
+        private const string ImmutableQueueGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableQueue`1";
+
+        private const string ImmutableSortedSetGenericTypeName = "System.Collections.Immutable.ImmutableSortedSet`1";
+
+        private const string ImmutableHashSetGenericTypeName = "System.Collections.Immutable.ImmutableHashSet`1";
+        private const string ImmutableSetGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableSet`1";
+
+        // This method is invoked with reflection in CreateFromList.
+        private static IEnumerable CreateFromListInternal<T>(MethodInfo createImmutableMethod, IList sourceList)
+        {
+            List<T> list = new List<T>();
+
+            foreach (object item in sourceList)
+            {
+                if (item is T itemT)
+                {
+                    list.Add(itemT);
+                }
+            }
+
+            return (IEnumerable)createImmutableMethod.Invoke(null, new object[] { list } );
+        }
+
+        public override IEnumerable CreateFromList(Type enumerableType, Type elementType, IList sourceList)
+        {
+            Debug.Assert(enumerableType.IsGenericType);
+
+            Type underlyingType = enumerableType.GetGenericTypeDefinition();
+            Type constructingType;
+            string name = underlyingType.FullName;
+
+            switch (name)
+            {
+                case ImmutableListGenericTypeName:
+                case ImmutableListGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableList);
+                    break;
+                case ImmutableArrayGenericTypeName:
+                    constructingType = typeof(ImmutableArray);
```

Comment:
I'm going to throw a fun little curve ball at you: ImmutableArray is a struct. You need to support serializing and deserializing a nullable ImmutableArray.

```cs
public class Foo
{
    public ImmutableArray<int>? Values { get; set; }
}
```

Add a test.

Isn't writing serializers *fun* 😬

Reply:
> Isn't writing serializers _fun_ 😬

Only a little. 😬

I've removed ImmutableArray support from this change. I'll follow up on a separate PR.